Frontend.XF: added the settings icon and side menu

kubaflo opened the pull request:
|iOS | Android |
|----|----|
| ![Simulator Screen Shot - iPhone 14 Pro Max - 2023-01-30 at 15 32 04](https://user-images.githubusercontent.com/42434498/215507370-ca64f823-c66b-4c8f-a44e-67f3386e2fa1.png)     |  ![Screenshot_1675089174](https://user-images.githubusercontent.com/42434498/215507430-b051849e-13cd-4876-a034-75171acc0fc4.png) |
|   ![Simulator Screen Shot - iPhone 14 Pro Max - 2023-01-30 at 15 32 07](https://user-images.githubusercontent.com/42434498/215507560-a7cc9d60-1b28-4674-976a-5f95e0c4ae13.png)   |  ![Screenshot_1675089176](https://user-images.githubusercontent.com/42434498/215507629-51ff4cd7-4a9d-4a27-8da2-a7fa687d9b4c.png)     |

knocte commented:
@kubaflo why do you include the two PNG logo files? there are already logo files in the iOS project

knocte commented:
> there are already logo files in the iOS project

Here: https://github.com/nblockchain/geewallet/tree/frontend/src/GWallet.Frontend.XF.iOS/Assets.xcassets/AppIcon.appiconset

knocte commented on src/GWallet.Frontend.XF/GWallet.Frontend.XF.fsproj:
```diff
@@ -141,5 +144,6 @@
     <PackageReference Include="Fsdk" Version="0.6.0--date20230118-0634.git-b829db2">
       <GeneratePathProperty></GeneratePathProperty>
     </PackageReference>
+    <EmbeddedResource Include="Resources\Fonts\FontAwesomeSolid.otf" />
```
@kubaflo let's remove this, we can improve the design in a separate PR

kubaflo replied:
Okay

knocte commented on src/GWallet.Frontend.XF/BalancesPage.xaml.fs:
```diff
@@ -535,3 +535,9 @@ type BalancesPage(state: FrontendHelpers.IGlobalAppState,
+        this.IsPresented <- true
+
+    member this.OpenSettings_Tapped(_: obj, _: EventArgs) =
+        Application.Current.MainPage.DisplayAlert("Settings", "Succesfully opened the settings page", "Ok") |> ignore
```
@kubaflo let's hook up the real feature, same as what Frontend.Console is doing, please

kubaflo replied:
But isn't it a request beyond the scope of this task? I thought I was supposed to add an icon and a side menu, not implement the things in the settings inside the app. I can do it but I don't think that it should be done in this PR

knocte replied:
There are things that can be split to other PRs and there are things that cannot. If I merge this PR as is, then the `frontend` branch would be in an unreleasable state, you understand? We cannot show to end-users certain UI that does nothing.

kubaflo replied:
@knocte okay, I will implement it

kubaflo replied:
@knocte can I use popups to implement this functionality or you'd prefer separate pages?

knocte replied:
it depends; if popups mean a lot of platform-specific code, I'd prefer pages

kubaflo commented:
> @kubaflo why do you include the two PNG logo files? there are already logo files in the iOS project

Fixed

knocte commented on src/GWallet.Frontend.XF/GWallet.Frontend.XF.fsproj:
```diff
@@ -141,5 +142,6 @@
     <PackageReference Include="Fsdk" Version="0.6.0--date20230118-0634.git-b829db2">
       <GeneratePathProperty></GeneratePathProperty>
     </PackageReference>
+    <EmbeddedResource Include="img\settings_80x80.png" />
```
let's rather place the image where other images are (see other EmbeddedResource elements in this .fsproj file)

knocte commented on src/GWallet.Frontend.XF/BalancesPage.xaml:
```diff
@@ -1,84 +1,138 @@
+<FlyoutPage xmlns="http://xamarin.com/schemas/2014/forms"
              xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
              xmlns:local="clr-namespace:GWallet.Frontend.XF"
+             xmlns:ios="clr-namespace:Xamarin.Forms.PlatformConfiguration.iOSSpecific;assembly=Xamarin.Forms.Core"
```
what's this iOS thing?

knocte replied:
what is a safe area?

kubaflo replied:
It's an area where the notch is. Here are the screens how the app would look like without it
![Simulator Screen Shot - iPhone 14 Pro Max - 2023-01-31 at 13 16 51](https://user-images.githubusercontent.com/42434498/215758778-e7f02a62-5a37-4cef-bdae-9e68c69f7d0b.png)
![Simulator Screen Shot - iPhone 14 Pro Max - 2023-01-31 at 13 16 55](https://user-images.githubusercontent.com/42434498/215758799-bdd052c0-0f15-4316-8c5c-e4487bc4c910.png)

knocte commented:
> This branch cannot be rebased due to conflicts

What happened here? (I generally recommend rebasing PRs via `git rebase` command, not `git merge`.)

kubaflo commented:
| iOS |![Simulator Screen Shot - iPhone 14 Pro Max - 2023-02-04 at 00 27 55](https://user-images.githubusercontent.com/42434498/216729802-001ab9d3-919a-4376-9462-e81e741b36ff.png) | ![Simulator Screen Shot - iPhone 14 Pro Max - 2023-02-04 at 00 28 11](https://user-images.githubusercontent.com/42434498/216729849-b547df51-5b8c-443a-9af3-3b05fb392037.png) | ![Simulator Screen Shot - iPhone 14 Pro Max - 2023-02-04 at 00 28 34](https://user-images.githubusercontent.com/42434498/216729872-5a7568eb-ba9a-4d1d-8cf0-7f41b40b12b6.png) |
|----|---|---|---|
| Android | ![Screenshot_1675467182](https://user-images.githubusercontent.com/42434498/216730468-110ea2ff-484b-4878-ad36-686ded9a6453.png) | ![Screenshot_1675467488](https://user-images.githubusercontent.com/42434498/216730506-c484b38f-5204-4853-b31b-82e9b0a211c6.png) | ![Screenshot_1675467478](https://user-images.githubusercontent.com/42434498/216730524-31347f82-b660-413d-8d92-a0452b6cdb8e.png)

@knocte This is how it looks now :)

kubaflo commented:
@knocte I did the rebase and removed colours

knocte commented:
> @knocte I did the rebase,

And you also squashed commits! Nice.

But CI is red because of this warning:

```
geewallet/src/GWallet.Frontend.XF/SettingsPage.xaml.fs(58,17): error FS0020: The result of this expression has type 'Async<unit>' and is implicitly ignored. Consider using 'ignore' to discard this value explicitly, e.g. 'expr |> ignore', or 'let' to bind the result to a name, e.g. 'let result = expr'. [/__w/geewallet/geewallet/src/GWallet.Frontend.XF/GWallet.Frontend.XF.fsproj]
```

We have warnings as errors in Release mode :)

kubaflo commented:
> > @knocte I did the rebase,
> 
> And you also squashed commits! Nice.
> 
> But CI is red because of this warning:
> 
> ```
> geewallet/src/GWallet.Frontend.XF/SettingsPage.xaml.fs(58,17): error FS0020: The result of this expression has type 'Async<unit>' and is implicitly ignored. Consider using 'ignore' to discard this value explicitly, e.g. 'expr |> ignore', or 'let' to bind the result to a name, e.g. 'let result = expr'. [/__w/geewallet/geewallet/src/GWallet.Frontend.XF/GWallet.Frontend.XF.fsproj]
> ```
> 
> We have warnings as errors in Release mode :)

I fixed this error. However, if it CI after this is still red, I will need your help :)

knocte commented on src/GWallet.Frontend.XF/SettingsPage.xaml:
```diff
@@ -0,0 +1,125 @@
+    xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
+    x:Class="GWallet.Frontend.XF.SettingsPage">
+
+    <AbsoluteLayout>
```
@kubaflo question, is there any specific reason of why you used AbsoluteLayout here instead of an non-absolute alternative?

kubaflo replied:
Yes, it is necessary as I want the 'mainLayout' to occupy 70% of the page's width
```
<StackLayout x:Name="mainLayout"
                      AbsoluteLayout.LayoutFlags="XProportional,SizeProportional"
                      AbsoluteLayout.LayoutBounds="0.5,50,0.7,1"
                      VerticalOptions="FillAndExpand">```

knocte commented on src/GWallet.Frontend.XF/BalancesPage.xaml:
```diff
@@ -1,84 +1,143 @@
+                <Image HeightRequest="80"
+                       WidthRequest="80"
+                       Margin="0,0,0,10"
+                       x:Name="logoInSidemenu"/>
```
@kubaflo cosmetic: let's rename this to sideMenuImg (this way, in case we decide to stop using a logo in the future, we don't need to rename it again)

knocte commented on src/GWallet.Frontend.XF/BalancesPage.xaml:
```diff
@@ -1,84 +1,143 @@
+                    <Label VerticalOptions="Center" Text="#"/>
+                    <Label Text="Check you still remember your secret recovery phrase"/>
+                    <StackLayout.GestureRecognizers>
+                        <TapGestureRecognizer CommandParameter="Check you still remember your secret recovery phrase" Tapped="OpenSettings_Tapped"/>
```
@kubaflo same 2 cosmetic things here ^

knocte commented on src/GWallet.Frontend.XF/BalancesPage.xaml:
```diff
@@ -1,84 +1,143 @@
+                    <Label VerticalOptions="Center" Text="#"/>
+                    <Label Text="Wipe your current wallet, in order to start from scratch"/>
+                    <StackLayout.GestureRecognizers>
+                        <TapGestureRecognizer CommandParameter="Wipe your current wallet, in order to start from scratch" Tapped="OpenSettings_Tapped"/>
```
@kubaflo same 2 cosmetic things here ^

knocte commented on src/GWallet.Frontend.XF/BalancesPage.xaml:
```diff
@@ -1,84 +1,143 @@
-            <controls:CircleChartView x:Name="normalChartView"
-                                     HorizontalOptions="FillAndExpand"
-                                     VerticalOptions="FillAndExpand"/>
+                <controls:CircleChartView x:Name="normalChartView"
```
@kubaflo (this way, this part will not show up in the diff)

knocte commented on src/GWallet.Frontend.XF/BalancesPage.xaml.fs:
```diff
@@ -46,6 +46,8 @@ type BalancesPage(state: FrontendHelpers.IGlobalAppState,
     let contentLayout = base.FindByName<StackLayout> "contentLayout"
     let normalChartView = base.FindByName<CircleChartView> "normalChartView"
     let readonlyChartView = base.FindByName<CircleChartView> "readonlyChartView"
+    let logoInSideMenu = base.FindByName<Image> "logoInSidemenu"
```
@kubaflo as we decided to rename the image name to `sideMenuImg`, let's also change the variable name here ^

knocte commented on src/GWallet.Frontend.XF/BalancesPage.xaml.fs:
```diff
@@ -46,6 +46,8 @@ type BalancesPage(state: FrontendHelpers.IGlobalAppState,
     let normalChartView = base.FindByName<CircleChartView> "normalChartView"
     let readonlyChartView = base.FindByName<CircleChartView> "readonlyChartView"
+    let logoInSideMenu = base.FindByName<Image> "logoInSidemenu"
+    let settingsButton = base.FindByName<ImageButton> "settingsIcon"
```
@kubaflo all widgets declared above with `base.FindByName<'T>` are using the same variable name as the name of the widget; let's do the same here, and given that the type is ImageButton, let's finish the variable/widget name with `ImgBtn`, for example how about `settingsBurgerLikeImgBtn`?

knocte commented on src/GWallet.Frontend.XF/BalancesPage.xaml.fs:
```diff
@@ -535,3 +539,11 @@ type BalancesPage(state: FrontendHelpers.IGlobalAppState,
             this.CancelBalanceRefreshJobs()
         )
+
+    member this.OpenFlyout_Clicked(_: obj, _: EventArgs) =
```
@kubaflo cosmetic:

- let's use `self` instead of `this`
- let's use `_sender` and `_args` for the arguments

knocte replied:
oh, and let's not use underscore in the name ^

knocte commented on src/GWallet.Frontend.XF/BalancesPage.xaml.fs:
```diff
@@ -535,3 +539,11 @@ type BalancesPage(state: FrontendHelpers.IGlobalAppState,
+    member this.OpenFlyout_Clicked(_: obj, _: EventArgs) =
+        this.IsPresented <- true
+
+    member this.OpenSettings_Tapped(_: obj, e: EventArgs) =
```
@kubaflo cosmetic:

- let's use `self` instead of `this`
- let's use `_sender` and `evArgs` for the arguments
- let's not use underscore in the name.

knocte commented on src/GWallet.Frontend.XF/BalancesPage.xaml.fs:
```diff
@@ -535,3 +539,11 @@ type BalancesPage(state: FrontendHelpers.IGlobalAppState,
+
+    member this.OpenSettings_Tapped(_: obj, e: EventArgs) =
+        let title = ((e :?> TappedEventArgs).Parameter :?> string)
+        let settingsPage () = SettingsPage(title) :> Page
```
@kubaflo cosmetic: no need for parentheses in the line above

knocte commented on src/GWallet.Frontend.XF/SettingsPage.xaml.fs:
```diff
@@ -0,0 +1,68 @@
+
+        } |> Async.StartImmediate
+
+    member this.datePicker_DateSelected (_: obj) (e: Xamarin.Forms.DateChangedEventArgs) =
```
@kubaflo cosmetic, please rename to _sender and `evArgs`

knocte commented on src/GWallet.Frontend.XF/SettingsPage.xaml.fs:
```diff
@@ -0,0 +1,68 @@
+
+    //loading & result
+    let resultMessage = base.FindByName<Label> "resultMessage"
+    let loadingIndicator= base.FindByName<ActivityIndicator> "loadingIndicator"
```
@kubaflo cosmetic, put space before `=` (I know this little things are annoying to fix, that's why we plan to use an automatic code formatter soon)

knocte commented on src/GWallet.Frontend.XF/SettingsPage.xaml.fs:
```diff
@@ -0,0 +1,68 @@
+    //loading & result
+    let resultMessage = base.FindByName<Label> "resultMessage"
+    let loadingIndicator = base.FindByName<ActivityIndicator> "loadingIndicator"
+    let option = option;
```
@kubaflo no need for semicolon here ^

knocte commented on src/GWallet.Frontend.XF/SettingsPage.xaml.fs:
```diff
@@ -0,0 +1,68 @@
+        this.Init()
+                    
+    member self.Init () =
+        titleLabel.Text<-option;
```
@kubaflo no need for semicolon here ^ (and please add spaces before and after the `<-` operator)

knocte commented on src/GWallet.Frontend.XF/SettingsPage.xaml.fs:
```diff
@@ -0,0 +1,68 @@
+
+open GWallet.Backend
+
+type SettingsPage(option:string) as this =
```
@kubaflo cosmetic: add space after colon

knocte commented:
Last cosmetic thing: you titled this PR properly, with a "Frontend.XF" as area, but forgot to do the same for the commit msg :)

knocte commented:
> Frontend.XF: Added the side menu and settings page.
> Frontend.XF: Added the settings page.
> Frontend.XF: Added the settings icon and side menu.

Frontend.XF is only supposed to be written once, just in the title of the commit message.

knocte commented on src/GWallet.Frontend.XF/GWallet.Frontend.XF.fsproj:
```diff
@@ -102,6 +105,10 @@
       <DependentUpon>SendPage.xaml</DependentUpon>
       <SubType>Code</SubType>
+    </Compile>
+        <Compile Include="SettingsPage.xaml.fs">
```
@kubaflo cosmetic ^ see above bad indentation

knocte commented on src/GWallet.Frontend.XF/SettingsPage.xaml.fs:
```diff
@@ -0,0 +1,68 @@
+
+open GWallet.Backend
+
+type SettingsPage(option : string) as this =
```
@kubaflo I said after colon, before the colon shouldn't have a space sorry

knocte commented on src/GWallet.Frontend.XF/SettingsPage.xaml.fs:
```diff
@@ -0,0 +1,68 @@
+
+        } |> Async.StartImmediate
+
+    member self.OnDatePickerDateSelected (_sender: obj) (evArgs: Xamarin.Forms.DateChangedEventArgs) =
```
@kubaflo no need for Xamarin.Forms here, as the ns is already opened

knocte commented on src/GWallet.Frontend.XF/SettingsPage.xaml.fs:
```diff
@@ -0,0 +1,68 @@
+    //loading & result
+    let resultMessage = base.FindByName<Label> "resultMessage"
+    let loadingIndicator = base.FindByName<ActivityIndicator> "loadingIndicator"
+    let option = option
```
@kubaflo this line above is not needed

knocte commented on src/GWallet.Frontend.XF/BalancesPage.xaml:
```diff
@@ -1,74 +1,131 @@
-                                     VerticalOptions="FillAndExpand"/>
+                                        IsVisible="False"
+                                        HorizontalOptions="FillAndExpand"
+                                        VerticalOptions="FillAndExpand"/>
```
@kubaflo see above ^ this space additions are not needed

kubaflo commented:
> We have fixed the check-password performance problem, if you rebase this PR then we can test the improvement with the settings page.

It works!!!

knocte commented:
The closing of this PR was unintended (it happened because the 'frontend' branch has now disappeared). @aarani can you please re-create this PR as it was, but targetting the master branch now? Thanks

aarani commented:
Superseded by #233
